consensus/hotstuff/signature.go

```diff
@@ -0,0 +1,57 @@
+package hotstuff
+
+import (
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type RandomBeaconReconstructor interface {
+	// Add a share
+	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+	HasSufficientShares() bool
+	// assume it has sufficient shares
+	Reconstruct() (crypto.Signature, error)
+}
+
+// ThresholdSigAggregator aggregates the threshold signatures
+type ThresholdSigAggregator interface {
+	// TrustedAdd adds an already verified signature.
+	// return (true, nil) means the signature has been added
+	// return (false, nil) means the signature is a duplication
+	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+	// Aggregate assumes enough shares have been collected, it aggregates the signatures
+	// and return the aggregated signature.
+	// if called concurrently, only one threshold will be running the aggregation.
+	Aggregate() ([]byte, error)
+}
+
+// StakingSigAggregator aggregates the staking signatures
+type StakingSigAggregator interface {
+	// TrustedAdd adds an already verified signature.
+	// return (true, nil) means the signature has been added
+	// return (false, nil) means the signature is a duplication
+	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+
+	// Aggregate assumes enough shares have been collected, it aggregates the signatures
+	// and return the aggregated signature.
+	// if called concurrently, only one threshold will be running the aggregation.
+	Aggregate() ([]byte, error)
+}
+
+type Packer interface {
```

Comment:
I think this one can be implemented as part of QC builder object, I don't see too much value having an interface for combining and splitting, utility class that does this seems enough

Reply:
The reason I extracted this is:
1. These functions are stateless pure functions. Easier to write unit test to check the correctness at this low level.
2. Yes, the creation of the QC can be implemented in QC builder, which is a stateful object. But the validation for QC is stateless. That's why I extracted this encoding part, so that the `Combine` method can be used in stateful creation, and the `Split` can be used in stateless validation.

Reply:
Technically, we only need custom struct for consensus cluster, because there are staking sig and random beacon sig.

For collection cluster, we don't need custom struct, because the proposer sig is alway just a staking sig. 

And we currently only have consensus cluster and collection cluster.